atlasdb-autobatch/src/main/java/com/palantir/atlasdb/autobatch/Autobatchers.java

```diff
@@ -79,25 +94,89 @@ private Autobatchers() {}
+            Consumer<List<BatchElement<I, O>>> batchFunction, EventHandlerParameters parameters) {
+        return parameters
+                .batchFunctionTimeoutContext()
+                .map(context -> wrapWithTimeout(batchFunction, parameters.safeLoggablePurpose(), context))
+                .orElse(batchFunction);
+    }
+
+    private static <I, O> CoalescingRequestFunction<I, O> maybeWrapWithTimeout(
+            CoalescingRequestFunction<I, O> coalescingFunction, EventHandlerParameters parameters) {
+        return parameters
+                .batchFunctionTimeoutContext()
+                .map(context -> wrapWithTimeout(coalescingFunction, parameters.safeLoggablePurpose(), context))
+                .orElse(coalescingFunction);
+    }
+
+    private static <I, O> Consumer<List<BatchElement<I, O>>> wrapWithTimeout(
+            Consumer<List<BatchElement<I, O>>> delegate,
+            String safeLoggablePurpose,
+            TimeoutOrchestrationContext context) {
+        TimeLimiter limiter = SimpleTimeLimiter.create(context.exclusiveExecutor());
+        return elements -> {
+            try {
+                limiter.runWithTimeout(() -> delegate.accept(elements), context.batchFunctionTimeout());
+            } catch (InterruptedException e) {
+                Thread.currentThread().interrupt();
+                throw new RuntimeException(e);
+            } catch (TimeoutException e) {
+                log.info(
+                        "Autobatcher timed out and request had to be interrupted",
+                        SafeArg.of("safeLoggablePurpose", safeLoggablePurpose),
+                        e);
+                throw new RuntimeException(e);
+            } catch (UncheckedExecutionException e) {
+                Throwables.throwIfUnchecked(e.getCause());
+                throw new RuntimeException(e.getCause());
+            }
+        };
+    }
+
+    private static <I, O> CoalescingRequestFunction<I, O> wrapWithTimeout(
+            CoalescingRequestFunction<I, O> delegate, String safeLoggablePurpose, TimeoutOrchestrationContext context) {
+        TimeLimiter limiter = SimpleTimeLimiter.create(context.exclusiveExecutor());
+        return elements -> {
+            try {
+                return limiter.callWithTimeout(() -> delegate.apply(elements), context.batchFunctionTimeout());
+            } catch (InterruptedException e) {
+                Thread.currentThread().interrupt();
+                throw new RuntimeException(e);
+            } catch (TimeoutException e) {
+                log.info(
+                        "Autobatcher timed out and request had to be interrupted",
+                        SafeArg.of("safeLoggablePurpose", safeLoggablePurpose),
+                        e);
+                throw new RuntimeException(e);
+            } catch (ExecutionException | UncheckedExecutionException e) {
+                Throwables.throwIfUnchecked(e.getCause());
```

Comment:
Will an `ExecutionException` even have an unchecked cause? See why you would still do it this way, but curious

Reply:
In this context no, as `SimpleTimeLimiter`'s contract is that it is used for checked exceptions (though in general, yes). I found this preferable to handling the conditional logic here explicitly, though happy to change if you prefer the alternative.